src/main/java/org/spectrumauctions/sats/core/bidlang/xor/CatsXOR.java

```diff
@@ -0,0 +1,278 @@
+
+        private XORValue<CATSLicense> handleNulls(CATSLicense first) throws NoSuchElementException {
+            if (acceptNulls) {
+                return null;
+            }
+            originalLicenseQueue.add(first); // Add this license to the original queue again
+            if (hasNext() && ++retries < MAX_RETRIES) return next();
+            else throw new NoSuchElementException("After " + retries + " retries, no other bundle was found " +
+                    "that was not identical to the original bundle bid and is valid in terms of budget and " +
+                    "min_resale_value constraints. \n" +
+                    "Most likely, there are either almost no licenses to choose from or the original bundle is very" +
+                    "small and highly valued, so that it's difficult to create another bundle that satisfies the" +
+                    "constraints. Try again (maybe with a higher number of goods) or use the the iterator that handles" +
+                    "this situation with null-values.");
+        }
+
+        private CATSLicense selectLicenseToAdd(Bundle<CATSLicense> bundle) {
+            if (uniRng.nextDouble() <= world.getJumpProbability()) {
+                if (goods.size() == bundle.size()) return null; // Prevent infinite loop if there is no other license
+                CATSLicense randomLicense;
+                do {
+                    Iterator<CATSLicense> iterator = goods.iterator();
+                    int index = uniRng.nextInt(goods.size());
+                    for (int i = 0; i < index; i++) {
+                        iterator.next();
+                    }
+                    randomLicense = iterator.next();
+                } while(bundle.contains(randomLicense));
+
+                return randomLicense;
+            } else {
+                WeightedRandomCollection<CATSLicense> neighbors = new WeightedRandomCollection<>(uniRng);
+                // Filter the licenses that are not contained yet in the bundle and where there exists an edge to one
+                // of the licenses in the bundle.
+                goods.stream().filter(l -> !bundle.contains(l) && edgeExists(l, bundle))
+                        .forEach(g -> {
+                            double positivePrivateValue = bidder.getPrivateValues().get(g.getId()).doubleValue() - minValue;
+                            neighbors.add(positivePrivateValue, g);
+                        });
+                if (neighbors.hasNext()) return neighbors.next();
+                else return null;
+            }
+        }
+
+        private boolean edgeExists(CATSLicense license, Bundle<CATSLicense> bundle) {
+            for (CATSLicense l : bundle) {
+                if (world.getGrid().isAdjacent(license.getVertex(), l.getVertex()))
+                    return true;
+            }
+            return false;
+        }
+
+        /**
+         * @see Iterator#remove()
+         */
+        @Override
+        public void remove() {
```

Comment:
Why do you override the equivalent default method?

Reply:
Very good question indeed. As a matter of fact, I copied the idea from the other iterators. I've cleaned them up as well now ;)